core/ir/riscv64/codec.py

```diff
@@ -0,0 +1,1064 @@
+
+
+@unique
+class Format(Enum):
+    # Uncompressed instructions
+    R = 1
+    R4 = 2
+    I = 3
+    S = 4
+    B = 5
+    U = 6
+    J = 7
+    # Only compressed instructions below
+    CR = 8
+    CI = 9
+    CSS = 10
+    CIW = 11
+    CL = 12
+    CS = 13
+    CA = 14
+    CB = 15
+    CJ = 16
+
+    def __str__(self) -> str:
+        return self.name.lower()
+
+
+class Instruction:
+    def __init__(self, name: str, fmt: str, mask: int, match: int, flds: List[str], ext: str) -> None:
+        self.name: str = name
+        self.fmt: Format = Format[fmt.upper()]
+        self.mask: int = mask
+        self.match: int = match
+        self.flds: List[Field] = [
+            Field.from_str(f) for f in flds if len(f) > 0]
+        self.ext = ext
+
+    def is_compressed(self) -> bool:
+        return (self.match & 0b11) < 0b11
+
+    def formatted_name(self) -> str:
+        '''
+        The formatted name will be used as the struct field name and enum variant.
+        '''
+        return self.name.lower().replace('.', "_")
+
+    def formatted_ext(self) -> str:
+        '''
+        The formatted extension name used as an enum variant.
+        '''
+        return f'RISCV64_ISA_EXT_{self.ext.upper()}'
+
+    def __str__(self) -> str:
+        fields = ' '.join([str(f) for f in self.flds])
+        sz = 15 if self.is_compressed() else 31
+        bits = ''.join([str((self.match >> i) & 1) if (
+            (self.mask >> i) & 1) == 1 else '.' for i in range(sz, -1, -1)])
+        return f"{self.name} | {self.fmt} | {fields} | {bits} | {self.ext}"
+
+
+class IslGenerator:
+    # Offset into opcode enum () to the first instruction. Updated by
+    # generate_opcodes().
+    OP_TBL_OFFSET = -1
+
+    class TrieNode:
+        def __init__(self, mask: int, shift: int, index: int, ctx: str = ''):
+            self.mask = mask
+            self.shift = shift
+            self.index = index
+            self.ctx = ctx
+
+    def __init__(self) -> None:
+        self.instructions: List[Instruction] = []
+        pass
+
+    def __fixup_compressed_inst(self, inst: Instruction):
+        opc = (inst.match & inst.mask) & 0x3
+        funct3 = (inst.match & inst.mask) >> 13
+        if (opc == 0 and funct3 not in [0, 0b100]):  # LOAD/STORE instructions
+            dbg(f'fixup: {inst.name} {[f.name for f in inst.flds]}')
+            # Immediate argument will handle the base+disp
+            inst.flds.pop(1)
+            dbg(f'    -> {" " * len(inst.name)} {[f.name for f in inst.flds]}')
+        elif (Field.CB_IMM in inst.flds):
+            # Compare-and-branch instructions need their branch operand moved
+            # to the 1st source operand slot as required by instr_set_target().
+            dbg(f'fixup: {inst.name} {[f.name for f in inst.flds]}')
+            cb_imm = inst.flds.pop(0)
+            inst.flds.append(cb_imm)
+            dbg(f'    -> {" " * len(inst.name)} {[f.name for f in inst.flds]}')
```

Comment:
Do we need an else clause to raise an exception or sth if  neither or these conditions are hit? Same below.

Reply:
No. The "else" case means that a given instruction doesn't need a fixup.